_algorithm_design/on_computing_next_greater_intervals_in_linear_time.md

```diff
@@ -0,0 +1,179 @@
+        raise NotImplementedError
+{% endhighlight %}
+
+Let's try to write an algorithm to compute the left bound for every element in the list using the observations and specified semantics.
+
+{% highlight python %}
+topo = MonotonicStack()
+for i, item in enumerate(list):
+    if topo.is_empty() or list[topo.get_next()] > item:
+        topo.push(i)
+    else:
+        while not topo.is_empty() && list[topo.get_next()] < item:
+            topo.pop()
+        if topo.is_empty():
+            leftBound[i] = -1
+        else:
+            leftBound[i] = topo.get_next()
+{% endhighlight %}
+
+### Analyzing the algorithmic complexity and reduction in search space
+
+So now we have an algorithm in place for maintaining the left bounds of elements in a list. Let's analyze the time and space complexity of this algorithm. The number of operations in the specified algorithm are proportional to the number of comparisons done. However, every comparison is followed by either a `push` or a `pop` call. Since every element gets pushed and popped out of stack exactly once, we get a total of $2N$ `push` or `pop` calls combined. Which also means $2N$ comparisons. Thus the time complexity of this algorithm is $O(N)$. Since the size of the utility stack can go upto a maximum of $N$, the space complexity is also $O(N)$.
+
+Answering the question of 'how much of the search space have we pruned?', we have pruned the search space from $O(N^{2})$ to $O(N)$.
+
+### Ok, but what about the rightBound?
+
+Consider the `pop` calls in the specified algorithm along with the element that is popped out. Clearly its left bound is the result of the `top` call but what is its right bound? Are you ready for the revelation? Hold tight onto your chairs and breathe deeply because this is going to blow your mind:
+.
+.
+.
+.
+`item`.
+
+By the condition on `item` in the algorithm, the execution will reach the `pop` call if and only if the invariant `list[topo.get_next()] < item` is satisfied. (assume that the stack is not empty in this case, otherwise the right bound index is N)
+
+And we're done. The left bound of the popped element is the result of the `top` call while the right bound is `item`. Since we know the left bound and right bound of an item when it is popped, it is best if we perform any `visit(element, leftBound, rightBound)` operations at this point. [NOTE: With appropriate design choice, one can write the visit routine with the given signature. : END OF NOTE]
+
+### Final algorithm
+
+{% highlight python %}
+topo = MonotonicStack()
+for i, item in enumerate(list):
+    if topo.is_empty() or list[topo.get_next()] > item:
+        topo.push(i)
+    else:
+        while not topo.is_empty() && list[topo.get_next()] < item:
+            element = topo.get_next()
+            topo.pop()
+            leftBound = topo.get_next() if not topo.is_empty() else -1
+            rightBound = i
+            visit(element, leftBound, rightBound)
+        topo.push(i)
```

Comment:
This can be rewritten as follows, 
```
        while not topo.is_empty() && list[topo.get_next()] < item:
            element = topo.get_next()
            topo.pop()
            leftBound = topo.get_next() if not topo.is_empty() else -1
            rightBound = i
            visit(element, leftBound, rightBound)
        topo.push(i)
```

`if - else` is redundant.

Reply:
Yes right! I've rewritten it like you've mentioned.